crates/flux-driver/src/collector/mod.rs

```diff
@@ -536,6 +542,102 @@ impl<'a, 'tcx> SpecCollector<'a, 'tcx> {
             self.specs.infer_opts.insert(def_id, infer_opts);
         }
     }
+
+    fn collect_detached_specs(&mut self, attrs: &mut FluxAttrs, def_id: LocalDefId) -> Result {
+        let Some(detached_specs) = attrs.detached_specs() else {
+            return Ok(());
+        };
+
+        self.collect_detached_specs_visitor(detached_specs, def_id)
+    }
+
+    fn collect_detached_specs_visitor(
+        &mut self,
+        detached_specs: surface::DetachedSpecs,
+        parent: LocalDefId,
+    ) -> Result {
+        let mut idents = detached_specs
+            .items
+            .into_iter()
+            .map(|item| (item.ident(), (item, None)))
+            .collect();
+
+        resolve_idents_in_scope(self.tcx, parent, &mut idents);
+
+        for (ident, (item, def_id)) in idents {
+            let Some(def_id) = def_id else {
+                return Err(self.errors.emit(errors::AttrMapErr {
+                    span: ident.span,
+                    message: format!("unresolved identifier `{ident}`"),
+                }));
+            };
+            let owner_id = self.tcx.local_def_id_to_hir_id(def_id).owner;
+            self.collect_detached_item(owner_id, item)?;
+        }
+        Ok(())
+    }
+
+    fn collect_detached_item(&mut self, owner_id: OwnerId, item: surface::DetachedItem) -> Result {
+        match item {
+            surface::DetachedItem::FnSig(_, fn_sig) => {
+                let spec = self
+                    .specs
+                    .fn_sigs
+                    .entry(owner_id)
+                    .or_insert(surface::FnSpec {
+                        fn_sig: None,
+                        qual_names: None,
+                        reveal_names: None,
+                    });
+
+                if spec.fn_sig.is_some() {
+                    let name = self.tcx.def_path_str(owner_id.to_def_id());
+                    return Err(self.errors.emit(errors::AttrMapErr {
+                        span: fn_sig.span,
+                        message: format!("multiple specs for `{name}`"),
+                    }));
+                }
+
+                spec.fn_sig = Some(*fn_sig);
+                Ok(())
+            }
+            surface::DetachedItem::Mod(_, detached_specs) => {
+                self.collect_detached_specs_visitor(detached_specs, owner_id.def_id)
+            }
+        }
+    }
+}
+
+struct DetachedIdentResolver<'a> {
+    items: &'a mut HashMap<Ident, (surface::DetachedItem, Option<LocalDefId>)>,
+}
+
+impl<'tcx> hir::intravisit::Visitor<'tcx> for DetachedIdentResolver<'_> {
+    fn visit_item(&mut self, item: &'tcx Item<'tcx>) {
```

Comment:
Oh I thought we were going to use [`TyCtxt::module_chilren_local`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TyCtxt.html#method.module_children_local) like we do [in name resolution](https://github.com/flux-rs/flux/blob/8ff117f14ed85215256bb4f60961abc06ccbbe49/crates/flux-desugar/src/resolver.rs#L693). This api is, in my opinion, slightly nicer to use because you don't need the visitor. It's also a bit lower level so we could eventually reuse it for writing extern specs.

Reply:
Did we decide to make this change to module-children or keep the hir-visitor?

Reply:
Hmm there is something odd. 

Note that we want to traverse the `ModuleChildren` recursively (for nested `mod`).

The odd thing is that `ModChild` only has the `DefId` of the inner definitions, not the `LocalDefId`.

However to (recursively) get the children, we need the `LocalDefId`? 

I suppose we could use [`as_local`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_hir/def_id/struct.DefId.html#method.as_local)

Is that what you had in mind?

Reply:
ok, have switched to the `module_children_local` (should be easy enough to revert if we want...)